python/ray/train/torch/xla/config.py

```diff
@@ -0,0 +1,169 @@
+        )
+        if os.path.exists(parallel_compile_workdir):
+            shutil.rmtree(parallel_compile_workdir)
+        os.makedirs(parallel_compile_workdir, exist_ok=True)
+
+        # Users can set the cache directory using --cache_dir in NEURON_CC_FLAGS or by
+        # using NEURON_COMPILE_CACHE_URL. --cache_dir takes precedence.
+        explicit_cache_dir = None
+        if neuron_cc_flags := os.environ.get("NEURON_CC_FLAGS"):
+            if s := re.search(r"--cache_dir[= ](\S+)", neuron_cc_flags):
+                explicit_cache_dir = s.group(1)
+
+        parallel_compile(
+            parallel_compile_workdir,
+            CacheUrl.get_cache_url(explicit_cache_dir),
+        )
+
+
+class _TorchAwsNeuronXLABackend(Backend):
+    unique_run_id: str = str(uuid.uuid4())
+
+    def on_start(self, worker_group: WorkerGroup, backend_config: TorchXLAConfig):
+        """Logic ran right before training is started."""
+
+        # On previous worker failure, we don't run graceful shutdown on workers.
+        # This would leak any running xrt server.
+        worker_group.execute(_kill_xrt_server)
+
+        # Get master address and port from the first worker.
+        master_addr, master_port = worker_group.execute_single(0, get_address_and_port)
+
+        def set_env_vars(addr, port):
+            os.environ["MASTER_ADDR"] = addr
+            os.environ["MASTER_PORT"] = str(port)
+            # To trigger the xrt server
+            os.environ["TORCHELASTIC_RUN_ID"] = self.unique_run_id
+
+        # Set the env vars on all workers.
+        worker_group.execute(set_env_vars, addr=master_addr, port=master_port)
+
+        # Set up env vars for neuron parallel compilation graph extraction
+        if backend_config.neuron_parallel_compile:
+            logger.info("Extracting graphs for Neuron parallel compilation")
+            worker_group.execute(_set_neuron_parallel_compile_env_vars)
+
+    def on_training_start(
+        self, worker_group: WorkerGroup, backend_config: TorchXLAConfig
+    ):
+        """
+        Configure the environment variables for the worker group.
+        And initialize the xla distributed process group.
+        TODO: Current setup only supports homogenous cluster with
+         neuron_cores accelerator and xrt runtime.
+        """
+        worker_group.execute(_set_xla_env_vars)
+        worker_group.execute(_setup_xla_torch_process_group)
+
+    def on_shutdown(self, worker_group: WorkerGroup, backend_config: TorchXLAConfig):
+        """
+        Logic ran right after training is finished.
+        This is a sanity cleanup to kill xrt server, and to optionally
+        run neuron parallel graph compilation
+        """
+        worker_group.execute(_kill_xrt_server)
+
+        # Compile the extracted graphs. This must run at end of training.
+        if backend_config.neuron_parallel_compile:
+            worker_group.execute(_neuron_compile_extracted_graphs)
```

Comment:
Why are we compiling the graph at the end? How would the people use the compiled graph?

Reply:
This code only runs during a pre-compilation step which happens when neuron_parallel_compile is set to true. `_neuron_compile_extracted_graphs()` must run at the end of the job (after a short number of training iterations) when all the graphs have been encountered.

See: [1](https://github.com/5cp/data-on-eks/tree/kuberay_trn1/ai-ml/trainium-inferentia/examples/kuberay_llama2_pretrain#run-precompilation-job) and [2](https://github.com/5cp/data-on-eks/blob/kuberay_trn1/ai-ml/trainium-inferentia/examples/kuberay_llama2_pretrain/docker/llama2_pretrain/launch_precompile_job.sh#L3) and [3](https://github.com/5cp/data-on-eks/blob/kuberay_trn1/ai-ml/trainium-inferentia/examples/kuberay_llama2_pretrain/docker/llama2_pretrain/ray_train_llama2.py#L198).

After precompilation, the user simply runs without neuron_parallel_compile to use the cached graphs and this avoids `_neuron_compile_extracted_graphs()`

Reply:
@woshiyyya does the above make sense?